src/universal/mutations/CreateReflectionMutation.js

```diff
@@ -0,0 +1,113 @@
+/**
+ * Creates a reflection for the retrospective meeting.
+ *
+ * @flow
+ */
+import {commitMutation} from 'react-relay';
+import {Environment, RecordProxy, RecordSourceSelectorProxy} from 'relay-runtime';
+
+import clientTempId from 'universal/utils/relay/clientTempId';
+import makeEmptyStr from 'universal/utils/draftjs/makeEmptyStr';
+
+type Variables = {
+  content?: string,
+  meetingId: string,
+  retroPhaseItemId: string,
+  sortOrder: number
+};
+
+type CompletedHandler = (response: ?Object, errors: ?Array<Error>) => void;
+
+type ErrorHandler = (error: Error) => void;
+
+graphql`
+  fragment CreateReflectionMutation_team on CreateReflectionPayload {
+    meeting {
+      id
+    }
+    reflection {
+      content
+      id
+      isViewerCreator
+      retroPhaseItemId
+      sortOrder
+    }
+}
+`;
+
+const mutation = graphql`
+  mutation CreateReflectionMutation($content: String, $meetingId: ID!, $retroPhaseItemId: ID!, $sortOrder: Float!) {
+    createReflection(content: $content, meetingId: $meetingId, retroPhaseItemId: $retroPhaseItemId, sortOrder: $sortOrder) {
+      ...CreateReflectionMutation_team @relay(mask: false)
+    }
+  }
+`;
+
+export const createReflectionUpdater = (payload: RecordProxy, store: RecordSourceSelectorProxy) => {
+  // TODO - https://github.com/dan-f/maeby
+  if (!payload) {
+    return;
+  }
+  const payloadMeeting = payload.getLinkedRecord('meeting');
+  if (!payloadMeeting) {
+    return;
+  }
+  const meetingId = payloadMeeting.getValue('id');
+  const reflection = payload.getLinkedRecord('reflection');
+  if (!reflection) {
+    return;
+  }
+  const meeting = store.get(meetingId);
+  if (!meeting) {
+    return;
+  }
+  const reflections = meeting.getLinkedRecords('reflections');
+  if (!reflections) {
+    return;
+  }
+  if (reflections.find((r) => r.getValue('id') === reflection.getValue('id'))) {
+    return;
+  }
+  const newReflections = [...reflections, reflection].sort((a, b) => {
+    const sortOrderA = a.getValue('sortOrder');
+    const sortOrderB = b.getValue('sortOrder');
+    return sortOrderA - sortOrderB;
+  });
+  meeting.setLinkedRecords(newReflections, 'reflections');
+};
+
+const getOptimisticResponse = (variables: Variables) => ({
```

Comment:
i'm not a huge for of `getOptimisticResponse` because it'll always turn into an `optimisticUpdater`.
For example, let's say a component calls `reflection.retroReflectionGroup.title`. 
You can't do that in a response. But in an updater, it's as easy as  `reflection.setLinkedRecord(store.get(retroReflectionGroup.getValue('id')), 'retroReflectionGroup')`

May also be nice to include other fields, such as the createdAt, updatedAt, etc.

Reply:
Assuming I understand you correctly, I'm not sure that I agree. When we get an actual server response, only the fields in the fragment on the payload are included. Our actual `updater` function is called with that response. Therefore, any side-effectual mutations to the store broader than the returned document itself must be implemented the same in the `updater` as in the `optimisticUpdater`, and it seems to me that the most straightforward way of achieving that is to have the same exact updater called with the as close to the same response as possible. From what I understand, the relay runtime calls the actual `updater` function with the optimistic response. It seems that the only case in which you'd want to do something differently in the optimistic updater vs the "real" updater is if you wanted to set some sort of "loading" or "mutation-in-progress" state.

Reply:
what you're leaving out is the magic that occurs inside relay before the `updater` gets called. Relay does all that magical linking up & updating records for you, so if you return a big enough piece of the graph, you'll never even need an updater function.

Reply:
Are you implying that if you include foreign keys in the response for a record, relay will link those together for you? Again, I wonder what value that brings us if if we didn't originally query on those links to begin with. I think I'm going to need further explanation on this one!